recommend/imageHandler.go

```diff
@@ -0,0 +1,366 @@
+	}
+	closeCheckErr(f, tarname)
+	log.WithFields(log.Fields{
+		"tar": tarname,
+	}).Info("dumped image to tar")
+	return tarname
+}
+
+func checkForSpec(spec string, fl []string) []string {
+	var matches []string
+	re := regexp.MustCompile(spec)
+	for _, name := range fl {
+		if re.Match([]byte(name)) {
+			matches = append(matches, name)
+		}
+	}
+	return matches
+}
+
+func getFileBytes(fname string) ([]byte, error) {
+	f, err := os.Open(filepath.Clean(fname))
+	if err != nil {
+		log.WithFields(log.Fields{
+			"file": fname,
+		}).Fatal("open file failed")
+	}
+	defer closeCheckErr(f, fname)
+	return io.ReadAll(f)
+}
+
+func (img *ImageInfo) readManifest(manifest string) {
+	// read manifest file
+	barr, err := getFileBytes(manifest)
+	if err != nil {
+		log.WithError(err).Fatal("manifest read failed")
+	}
+	var manres []map[string]interface{}
+	err = json.Unmarshal(barr, &manres)
+	if err != nil {
+		log.WithError(err).Fatal("manifest json unmarshal failed")
+	}
+	if len(manres) < 1 {
+		log.WithFields(log.Fields{
+			"len":     len(manres),
+			"results": manres,
+		}).Fatal("expecting atleast one config in manifest!")
+	}
+
+	var man map[string]interface{}
+	for _, man = range manres {
+		if man["RepoTags"] != nil {
+			break
+		}
+	}
+
+	// read config file
+	config := filepath.Join(tempDir, man["Config"].(string))
+	barr, err = getFileBytes(config)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"config": config,
+		}).Fatal("config read failed")
+	}
+	var cfgres map[string]interface{}
+	err = json.Unmarshal(barr, &cfgres)
+	if err != nil {
+		log.WithError(err).Fatal("config json unmarshal failed")
+	}
+	img.Arch = cfgres["architecture"].(string)
+	img.OS = cfgres["os"].(string)
+	for _, tag := range man["RepoTags"].([]interface{}) {
+		img.RepoTags = append(img.RepoTags, tag.(string))
+	}
+}
+
+type distroRule struct {
+	Distro string `json:"distro"`
+	Match  []struct {
+		Path string `json:"path"`
+	} `json:"match"`
+}
+
+//go:embed json/distro.json
+var distroJSON []byte
+
+var distroRules []distroRule
+
+func init() {
+	err := json.Unmarshal(distroJSON, &distroRules)
+	if err != nil {
+		color.Red("failed to unmarshal distro json rules")
+		log.WithError(err).Fatal("failed to unmarshal distro json rules")
+	}
+}
+
+func (img *ImageInfo) getDistro() {
+	for _, d := range distroRules {
+		match := true
+		for _, m := range d.Match {
+			matches := checkForSpec(filepath.Clean(tempDir+m.Path), img.FileList)
+			if len(matches) == 0 {
+				match = false
+				break
+			}
+		}
+		if len(d.Match) > 0 && match {
+			color.Green("Distribution %s", d.Distro)
+			img.Distro = d.Distro
+			return
+		}
+	}
+}
+
+func (img *ImageInfo) getImageInfo() {
+	matches := checkForSpec(filepath.Join(tempDir, "manifest.json"), img.FileList)
+	if len(matches) != 1 {
+		log.WithFields(log.Fields{
+			"len":     len(matches),
+			"matches": matches,
+		}).Fatal("expecting one manifest.json!")
+	}
+	img.readManifest(matches[0])
```

Comment:
What do you think about using https://pkg.go.dev/github.com/containers/image/manifest to read the manifest here?

Reply:
Awesome find. Let me check if I can use it.

Reply:
Hey @daemon1024 , using `containers/image` would be incredible i.e., it will allow us to work with any container registry, it will provide most of the other manifest API.

One problem that I saw with `containers/image` APIs is that it requires root access since it assumes that the containers are mounted as part of `/var/lib/containers`. ([ref](https://iximiuz.com/en/posts/working-with-container-images-in-go/))

Currently I am relying on docker friendly APIs and imo for the current version it should be good enough without requiring root privileges.

Reply:
Okay, let's keep it in mind for future then.